src/java/org/apache/cassandra/service/PendingRangeCalculatorService.java

```diff
@@ -46,47 +51,87 @@ public class PendingRangeCalculatorService
+
+    private final TokenMetadataProvider tokenMetadataProvider;
+
+    private void doUpdate()
     {
-        PendingRangeCalculatorServiceDiagnostics.taskStarted(1);
-        long start = currentTimeMillis();
-        Collection<String> keyspaces = Schema.instance.distributedKeyspaces().names().stream()
-                                                      .filter(keyspaceNamePredicate)
-                                                      .collect(Collectors.toList());
-
-        for (String keyspaceName : keyspaces)
-            calculatePendingRanges(Keyspace.open(keyspaceName).getReplicationStrategy(), keyspaceName);
-        if (logger.isTraceEnabled())
-            logger.trace("Finished PendingRangeTask for {} keyspaces in {}ms", keyspaces.size(), currentTimeMillis() - start);
-        PendingRangeCalculatorServiceDiagnostics.taskFinished();
+        // repeat until all keyspaced are consumed
+        while (!keyspacesWithPendingRanges.isEmpty())
+        {
+            long start = System.currentTimeMillis();
+
+            int updated = 0;
+            int total = 0;
+            PendingRangeCalculatorServiceDiagnostics.taskStarted(1);
+            try
+            {
+                Set<String> keyspaces = new HashSet<>(keyspacesWithPendingRanges);
+                total = keyspaces.size();
+                keyspacesWithPendingRanges.removeAll(keyspaces); // only remove those which were consumed
+
+                Iterator<String> it = keyspaces.iterator();
+                while (it.hasNext())
+                {
+                    String keyspaceName = it.next();
+                    try
+                    {
+                        calculatePendingRanges(keyspaceName);
+                        it.remove();
+                        updated++;
+                    }
+                    catch (RuntimeException | Error ex)
+                    {
+                        logger.error("Error calculating pending ranges for keyspace {}", keyspaceName, ex);
+                    }
+                }
+            }
+            finally
+            {
+                PendingRangeCalculatorServiceDiagnostics.taskFinished();
+                if (logger.isTraceEnabled())
+                    logger.trace("Finished PendingRangeTask for {} keyspaces out of {} in {}ms", updated, total, System.currentTimeMillis() - start);
+            }
+        }
     }
 
     public PendingRangeCalculatorService()
     {
+        this("PendingRangeCalculator", Schema.instance);
+    }
+
+    public PendingRangeCalculatorService(String executorName, Schema schema)
+    {
+        this(executorFactory().withJmxInternal()
+                              .configureSequential(executorName)
+                              .withRejectedExecutionHandler((r, e) -> {})  // silently handle rejected tasks, this::update takes care of bookkeeping
+                              .build(), TokenMetadataProvider.instance, schema);
```

Comment:
Nit: can the last two arguments be moved to separate line as in the original patch? It will improve readability, since all previous line are related for preparing the first argument to `this`.